src/core/shards/assert.cpp

```diff
@@ -92,9 +92,146 @@ struct IsNot : public Base {
+
+  void setParam(int index, const SHVar &inValue) {
+    switch (index) {
+    case 0:
+      destroyVar(_value);
+      cloneVar(_value, inValue);
+      break;
+    case 1:
+      _aborting = inValue.payload.boolValue;
+      break;
+    case 2:
+      _epsilon = inValue.payload.floatValue;
+      break;
+    default:
+      break;
+    }
+  }
+
+  SHVar getParam(int index) {
+    switch (index) {
+    case 0:
+      return _value;
+    case 1:
+      return Var(_aborting);
+    case 2:
+      return Var(_epsilon);
+    default:
+      return Var::Empty;
+    }
+  }
+
+  SHTypeInfo compose(const SHInstanceData &data) {
+    if (_value.valueType != data.inputType.basicType)
+      throw SHException("Input and value types must match.");
+
+    return data.inputType;
+  }
+
+  SHVar activate(SHContext *context, const SHVar &input) {
+    if (!isAlmost(input, _value)) {
+      SHLOG_ERROR("Failed assertion IsAlmost, input: {} expected: {}", input, _value);
+      if (_aborting)
+        abort();
+      else
+        throw ActivationError("Assert failed - IsAlmost");
+    }
+
+    return input;
+  }
+
+private:
+  bool isAlmost(const SHVar &lhs, const SHVar &rhs) {
+    if (lhs.valueType != rhs.valueType) {
+      return false;
+    }
+
+    switch (lhs.valueType) {
+    case SHType::Int:
+    case SHType::Int2:
+    case SHType::Int3:
+    case SHType::Int4:
+    case SHType::Int8:
+    case SHType::Int16:
+      return lhs == rhs;
+    case SHType::Float:
+      return isAlmost(lhs.payload.floatValue, rhs.payload.floatValue);
+    case SHType::Float2:
+      return (isAlmost(lhs.payload.float2Value[0], rhs.payload.float2Value[0]) &&
+              isAlmost(lhs.payload.float2Value[1], rhs.payload.float2Value[1]));
+    case SHType::Float3:
+      return (isAlmost(lhs.payload.float3Value[0], rhs.payload.float3Value[0]) &&
+              isAlmost(lhs.payload.float3Value[1], rhs.payload.float3Value[1]) &&
+              isAlmost(lhs.payload.float3Value[2], rhs.payload.float3Value[2]));
+    case SHType::Float4:
+      return (isAlmost(lhs.payload.float4Value[0], rhs.payload.float4Value[0]) &&
+              isAlmost(lhs.payload.float4Value[1], rhs.payload.float4Value[1]) &&
+              isAlmost(lhs.payload.float4Value[2], rhs.payload.float4Value[2]) &&
+              isAlmost(lhs.payload.float4Value[3], rhs.payload.float4Value[3]));
+    case SHType::Seq: {
+      if (lhs.payload.seqValue.len != rhs.payload.seqValue.len) {
+        return false;
+      }
+
+      auto almost = true;
+      for (uint32_t i = 0; i < lhs.payload.seqValue.len; i++) {
+        auto &suba = lhs.payload.seqValue.elements[i];
+        auto &subb = rhs.payload.seqValue.elements[i];
+        almost = almost && isAlmost(suba, subb);
+      }
+
+      return almost;
+    }
+    default:
+      return false;
+    }
+  }
+
+  bool isAlmost(double lhs, double rhs) { return __builtin_fabs(lhs - rhs) <= _epsilon; }
+
+  static inline Types MathTypes{{
+      CoreInfo::IntType,
```

Comment:
I would remove Int types cos if not ideally epsilon could be `2` and you'd have to implement the logic.

Reply:
I added Int because there was a sample with a sequence of numbers containing both integers and decimals.

Arguably, I could remove it from the list of supported types but still support it as fallback when from inside a sequence. In fact, it could just be described in the documentation that epsilon only affects decimal number and otherwise acts like `Is`.

Reply:
I could (should) also adda check that epsilon must be strictly between 0 and 1 (both ends excluded).